### PR TITLE
Enable nullable in UIActionEngine, fix NRE during the update operation

### LINKED ISSUE #12472: NRE raised when trying to collect the existing packages during an Update operation
1. Launch exp instance in debug mode
2. Set a breakpoint in https://github.com/NuGet/NuGet.Client/blob/f49ca00e27297fe7b4b82a2557d5b0caad907a06/src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs#L327.
3. Open a project with at least package that needs update.
4. CLick update
5. See breakpoint get hit.

The effect of this is that the telemetry does not get sent, but the try/catch hides it.

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/12472

Regression? Last working version:

## Description

Since Action can be null, 2. Set a breakpoint in https://github.com/NuGet/NuGet.Client/blob/f49ca00e27297fe7b4b82a2557d5b0caad907a06/src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs#L327 can throw an NRE. Once I fixed this NRE, I started hitting other ones.

In order to avoid any of these issues in the future, I've enabled nullable in the whole class, and addressed all the warnings.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [x] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs [hunk 1]
```diff
@@ -709,10 +707,15 @@ internal static List<TelemetryEvent> ToTelemetryPackageList(List<Tuple<string, s
...
         private async Task<bool> CheckPackageManagementFormatAsync(
-            INuGetProjectUpgraderService projectUpgrader,
+            INuGetProjectUpgraderService? projectUpgrader,
             INuGetUI uiService,
             CancellationToken cancellationToken)
         {
+            if (projectUpgrader == null)
+            {
+                return false;
+            } // todo nk?
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs]:
```suggestion
            }
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs [hunk 2]
```diff
@@ -1012,38 +1017,41 @@ private PackageLicenseInfo GeneratePackageLicenseInfo(IPackageSearchMetadata met
...
 
                 results.AddRange(completed);
 
                 if (completed.Length != allPackages.Length)
                 {
                     // get remaining package's metadata from remote repositories
-                    var remainingPackages = allPackages.Where(package => !completed.Any(pack => pack.Identity.Equals(package)));
-
-                    var remoteResults = (await TaskCombinators.ThrottledAsync(
+                    var remainingPackages = allPackages.Where(package => package != null && !completed.Any(pack => pack != null && pack.Identity.Equals(package)));
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs]: Naming?
```suggestion
                    var remainingPackages = allPackages.Where(package => package != null && !completed.Any(metadata=> metadata != null && metadata.Identity.Equals(package)));
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs [hunk 3]
```diff
@@ -1012,38 +1017,41 @@ private PackageLicenseInfo GeneratePackageLicenseInfo(IPackageSearchMetadata met
...
+#pragma warning restore CS8619 // Nullability of reference types in value doesn't match target type.
 
                     results.AddRange(remoteResults);
                 }
             }
             // check if missing metadata for any package
             if (allPackages.Length != results.Count)
             {
-                var package = allPackages.First(pkg => !results.Any(result => result.Identity.Equals(pkg)));
+                var package = allPackages.First(pkg => !results.Any(result => result != null && result.Identity.Equals(pkg)));
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs]: use explicit type

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs [hunk 4]
```diff
@@ -164,17 +165,15 @@ public async Task UpgradeNuGetProjectAsync(INuGetUI uiService, IProjectContextIn
             if (!string.IsNullOrEmpty(backupPath))
             {
                 string htmlLogFile = GenerateUpgradeReport(projectName, backupPath, upgradeInformationWindowModel);
-
-                Process process = null;
                 try
                 {
-                    process = Process.Start(htmlLogFile);
+                    var process = Process.Start(htmlLogFile);
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs]: use explicit type

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs]: I think this one is obvious.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs (hunk 2 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs]: I didn't change that, but I can fix it.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs (hunk 3 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs]: I didn't touch that, but I can fix it.

**Review by jeffkl — DISMISSED** [review]

**Review by erdembayar — DISMISSED** [review]

**Review by erdembayar — DISMISSED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs (hunk 4 above)
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs]:
```suggestion
                    using Process.Start(htmlLogFile);
```

or alternatively 

```suggestion
                    _ = Process.Start(htmlLogFile);
```

`Process` has native handles, so "normally" disposing is useful to allow the native resources to be released immediately, without having to wait for the GC to run the finalizer. I don't think it's relevant that the process is "fire and forget", so I expect that disposing it immediately should be fine.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs]: Fixed.

**Review by zivkan — APPROVED** [review]

**Pull request merged**